src/unstract/sdk/file_storage/fs_permanent.py

```diff
@@ -0,0 +1,63 @@
+from typing import Any, Union
+
+from unstract.sdk.exceptions import FileOperationError, FileStorageError
+from unstract.sdk.file_storage.constants import Common
+from unstract.sdk.file_storage.fs_impl import FileStorage
+from unstract.sdk.file_storage.fs_provider import FileStorageProvider
+
+
+class PermanentFileStorage(FileStorage):
+    SUPPORTED_FILE_STORAGE_TYPES = [
+        FileStorageProvider.GCS.value,
+        FileStorageProvider.S3.value,
+        FileStorageProvider.Azure.value,
+    ]
+
+    def __init__(
+        self,
+        provider: FileStorageProvider,
+        credentials: dict[str, Any] = {},
+    ):
+        if provider.value not in self.SUPPORTED_FILE_STORAGE_TYPES:
+            raise FileStorageError(
+                f"File storage provider is not supported in Permanent mode. "
+                f"Supported providers: {self.SUPPORTED_FILE_STORAGE_TYPES}"
+            )
+        if provider == FileStorageProvider.GCS:
+            super().__init__(provider, credentials)
+        else:
+            raise NotImplementedError
+
+    def copy_on_write(self, path1, path2):
+        self.fs.put(path1, path2)
```

Comment:
@gaya3-zipstack We could expose a `copy` API also in the interface, as that could be generic and stand alone too.

The lines 57-60 below collectively is actually the "copy on write" strategy.